### PR TITLE
Feature/9182 post filters rows ui

### LINKED ISSUE #9182: Update Post List row UI
As part of post filters project we'll be updating the post list cards a little bit. This includes a new action in `Restore Post` and an update to post status labels such as `Locally Changed`, `Private` etc. Here is the new design: https://cloudup.com/cBtFWQBCvZr. Note that this design is only the first iteration, so it might change before this issue is tackled.

### PR DESCRIPTION
**Pull request by malinajirka**
Partially fixes #9182  

**Note**
- The target branch should be updated to "feature/master-post-filters" when #9430 is merged.
- This PR doesn't try to refactor the code into the optimal state, it's just another iteration towards that goal.

**Changes**
- Extracts logic related to items in the PostList into a separate class - `PostListItemUiStateHelper`
- Updates the ui of post list items according to the current design (The retry button UI doesn't correspond to Zeplin - updating Zeplin is in Sylvester's backlog)

**TODO**
- implement restore and permanent delete actions
- only one label (eg. Local Draft, Uploading post etc.) is displayed at a time. It's an easy fix, I'm just not sure which labels are compatible.
- write unit tests

**To test:**
This PR touches all logic related to UI of items in the post list
- which labels are displayed (local draft, uploading, errors...)
- which actions are displayed (View, Submit/Publish/Retry/.., Edit, Trash/Delete ..)
- feature image, title, excerpt
- progress indicator, disabled overlay

Make sure clicks on actions are tracked and the actions work (don't forget to check the actions under "More" are also being tracked).

I would suggest testing only a couple of basic use cases and test the rest when we have all the unit tests + all the logic is implemented.

**Update release notes:**
- We'll update the release not when we merge the working branch into develop

### CODE REVIEW

**Comment by oguzkocer** [comment]
I am planning to review this PR on Mar 20, Wednesday as I want to spend some time thinking through things.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    val title: UiString?,
+    val excerpt: UiString?,
+    val imageUrl: String?,
+    val dateAndAuthor: UiString?,
+    val statusLabels: UiString?,
+    @ColorRes val statusLabelsColor: Int?,
+    val actions: List<PostListItemAction>,
+    val showProgress: Boolean,
+    val showOverlay: Boolean,
+    val onSelected: () -> Unit
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt]: :headbang: Thanks for catching this :X. I don't how I could forget about it.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                ivImageFeatured.visibility = View.VISIBLE
+                config.imageManager.load(ivImageFeatured, bmp)
+            } else {
+                ivImageFeatured.visibility = View.GONE
+                config.imageManager.cancelRequestAndClearImageView(ivImageFeatured)
+            }
+        }
+    }
+
+    private fun setTextOrHide(view: WPTextView, text: UiString?) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt]: :D yes, I have been planning to add it for some time now. I always feel like I should move all the usage in the code-base so it's not a good idea do it in a feature PR. However, I'll make sure to actually get to it next week.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        unhandledConflicts: Boolean,
+        capabilitiesToPublish: Boolean,
+        statsSupported: Boolean,
+        featuredImageUrl: String?,
+        formattedDate: String,
+        onAction: (PostModel, PostListButtonType, AnalyticsTracker.Stat) -> Unit
+    ): PostListItemUiState {
+        val postStatus: PostStatus = PostStatus.fromPost(post)
+
+        return PostListItemUiState(
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: Good call! Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9431/commits/3148b4d7ba9423f88ce178a6c3c8c90025229908

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        } else {
+            buttonTypes.add(PostListButtonType.BUTTON_DELETE)
+            buttonTypes.add(PostListButtonType.BUTTON_RESTORE)
+        }
+
+        if (canShowStats) {
+            buttonTypes.add(PostListButtonType.BUTTON_STATS)
+        }
+
+        return if (buttonTypes.size > MAX_NUMBER_OF_VISIBLE_ACTIONS) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: 👍 Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9431/commits/8ae0e6113bf05604a850d6709ccf5c4ef347fdb0

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    class SingleItem(buttonType: PostListButtonType, onButtonClicked: (PostListButtonType) -> Unit) :
+            PostListItemAction(buttonType, onButtonClicked)
+
+    class MoreItem(
+        val actions: List<PostListItemAction>,
+        onButtonClicked: (PostListButtonType) -> Unit
+    ) : PostListItemAction(PostListButtonType.BUTTON_MORE, onButtonClicked)
+}
+
+data class PostListItemUploadStatus(
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt]: 👍 Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9431/commits/335565e5ff1d7b533c8c27924ce98bbaf805467f

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt]: Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9431/commits/b7026392119ba68701380959c961366112562fe2

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+data class PostListItemUiState(
+    val remotePostId: Long,
+    val localPostId: Int,
+    val title: UiString?,
+    val excerpt: UiString?,
+    val imageUrl: String?,
+    val dateAndAuthor: UiString?,
+    val statusLabels: UiString?,
+    @ColorRes val statusLabelsColor: Int?,
+    val actions: List<PostListItemAction>,
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiState.kt]: Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9431/commits/b7026392119ba68701380959c961366112562fe2

**Comment by malinajirka** [comment]
Thank you so much for reviewing this! I'm really sorry for the size of the change set, but I felt we'd need to test all the scenarios before we merge the `feature/master-post-filters` into `develop` anyway so it's worth making the code a bit cleaner.

> I think our best course of action is to address a couple minor concerns I have, merge this PR and the subsequent few and go through the whole stack once it's settled. I think that'll be more comprehensive and time efficient.

I 100% agree, this is basically what I meant by "I would suggest testing only a couple of basic use cases and test the rest when we have all the unit tests + all the logic is implemented."

I'm sorry for introducing the diffing issue.. Tbh I have no idea how I could have forgotten. I can't wrap my head around it. 😕 

It's ready for another round. Thanks @oguzkocer

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt]: It's in my must-do todo list :)!

**Comment by oguzkocer** [comment]
> I'm really sorry for the size of the change set, but I felt we'd need to test all the scenarios before we merge the `feature/master-post-filters` into `develop` anyway so it's worth making the code a bit cleaner.

No problem at all, just to be clear, I agree with your approach and I definitely wasn't complaining about the size of the PR.

> ... this is basically what I meant by ...

Glad we are on the same page :)

> I'm sorry for introducing the diffing issue.. Tbh I have no idea how I could have forgotten. I can't wrap my head around it.

No worries, it happens, lots of moving pieces in this PR and in post list in general.

> It's ready for another round.

I think everything looks good, I'll do just a quick test and a very quick glance over again and merge this. 👍

**Pull request merged**